nixos: replace 'text' with structured PAM rules

Majiir opened the pull request:
Replaces most of the nixpkgs usages of the `text` option to define PAM service rules with the new `rules` option introduced in https://github.com/NixOS/nixpkgs/pull/255547. (There's one more `text` usage which is removed in https://github.com/NixOS/nixpkgs/pull/420885.) See commit descriptions for details.

This incidentally fixes a bug where the `gdm-fingerprint` and `vsftpd` PAM services are created with an invalid set of default rules when some (but not all) of the conditions are met to enable them.

How this fits into the PAM refactoring plan:

1. https://github.com/NixOS/nixpkgs/pull/255547
2. https://github.com/NixOS/nixpkgs/pull/420889 **:arrow_left: you are here**
3. refine the `rules` interface, using these in-tree rules as test cases
4. deprecate the `text` option
5. decompose `pam.nix` into modules that contribute `rules`

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
    - [x] `pam-file-contents`
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Majiir committed:
nixos/pam: add useDefaultRules option

This option is enabled by default to preserve the current behavior when
a new service is declared. Users may disable this option to more easily
create a service without any rules. In nixpkgs, we can use this option
to eliminate usage of the 'text' option where the entire service rule
stack is replaced.

Majiir committed:
nixos/pam: extract autoOrderRules to utils

This function is used to convert an ordered list of rules into an
attrset of rules with reasonable 'order' values. This reduces
boilerplate to define 'order' and makes it simple to switch how ordering
is managed in the future.

Majiir committed:
nixos: use full path to PAM modules

PAM rules with non-absolute module paths are rejected when apparmor is
used. In general, it helps (aside from readability) for all the module
paths to be absolute, especially when the user overrides the PAM
package.

Majiir committed:
nixos: replace 'text' with structured PAM rules

Several modules define whole PAM service rule stacks by overwriting the
default value of the 'text' option. Instead, we disable useDefaultRules
for these services and declare a new set of rules using the 'rules'
option. This option is considered experimental and hidden from users,
but it is supported for use within nixpkgs.

nyabinary commented:
Merge conflict :p

Majiir committed:
nixos/pam: add assertion for autoOrderRules

l0b0 commented:
Thank you, replacing strings with structured data is always welcome. But this is a rather large change, so it's hard to review. Would it be possible to include tests beyond `pam-file-contents` to verify that there's been no change of _any_ of the rules from the target branch? Or could you provide commands to easily compare the complete before and after?

Majiir commented:
I have pushed a second branch ([`pam_use_rules_option_explained`](https://github.com/NixOS/nixpkgs/compare/master...Majiir:nixpkgs:pam_use_rules_option_explained)) which breaks apart the last commit into more easily reviewed steps. Some of these steps are performed by regex replacement, so you can inspect the regex or a sample of the diff and have confidence in the whole commit. Other changes were done manually, but these are easier to review by hand. The two branches are identical ([diff](https://github.com/Majiir/nixpkgs/compare/pam_use_rules_option..pam_use_rules_option_explained)). The smaller steps are not always valid states, but it should show the transformation clearly.

For comparing the generated files, I often do something like this:

```
git checkout master
nixos-rebuild build
mv result result-old
git checkout with-pam-patch
nixos-rebuild build
diff -r result-old/etc/pam.d result/etc/pam.d
```

Some changes are expected, since the structured rules produce PAM files with a different format than the manual `text`. But aside from whitespace, comments and module paths, the rules should be identical.

l0b0 reviewed (approved):
I've only skimmed the rule translations, but this looks good. IIUC the main _functional_ change is pulling `autoOrderRules` out verbatim.

The `autoOrderRules` implementation is a bit hard to understand with the combination of `flip` and `pipe`, but that was already in the original code.

l0b0 commented:
I've taken the liberty to [link this in Matrix](https://matrix.to/#/!zihVdbsMjkhRfMQMXv:nixos.org/$sGqyUFn4XNNEv1tXxKUlwbp75bGt7PtJk1xwT3zbLSI?via=nixos.org&via=matrix.org&via=tchncs.de).

phaer reviewed (approved):
Rules look semantically identical, other changes sensible.

@Majiir This needs a rebase, but I think otherwise we could merge? :)

(Nit: I think one commit per module might have been both easier to review and to to revert in case thats necessary)

Majiir commented:
- Rebased.
- Re-ran the regex procedure, which added changes for `plasma-login-manager`.
- Added some parens so that `nix fmt` is no longer needed, which should reduce the frequency of merge conflicts.

zacharyarnaise reviewed (approved):
LGTM

phaer reviewed (approved):
After my code-only review last time, I rebuilt all 3 desktop/laptop and 6 server machines that include gdm, fprintd, and vsftpd at least. Once with nixos-unstable and once with this branch rebased on it and diffed all of /etc/pam.d.

All changes were
* whitespace
* using absolute store paths for pam modules.
* autorder comments.

Except for gdm-fingerprint where the file is now absent if fprintd is disabled, which makes sense. from the diff (`gdm-fingerprint` vs `gdm-fingerprint.text` in the `mkIf`).

Also ran the following pam-related vm-tests for good measure: 
```
nom-build --keep-going \
     -A nixosTests.pam-file-contents \
     -A nixosTests.pam-lastlog \
     -A nixosTests.pam-oath-login \
     -A nixosTests.pam-pgsql \
     -A nixosTests.pam-u2f \
     -A nixosTests.pam-u2f-polkit \
     -A nixosTests.pam-ussh \
     -A nixosTests.pam-zfs-key \
     -A nixosTests.slurm-pam \
     -A nixosTests.ssh-agent-auth \
     -A nixosTests.google-oslogin \
     -A nixosTests.auth-mysql \
     -A nixosTests.sssd-ldap \
     -A nixosTests.sssd-legacy-config \
     -A nixosTests.kanidm \
     -A nixosTests.fscrypt \
     -A nixosTests.systemd-homed \
     -A nixosTests.openssh
```

`google-oslogin` and `pam-zfs-key` failed, but they did so on my merge base (4c1018dae018162ec878d42fec712642d214fdfa) as well. 

Looks safe, fingers crossed :crossed_fingers:

catuhana commented:
The commit 4b864991aa141958fc701e4925527fbf45ff1bf0 (which is a part of this PR) gives an evaluation error when AppArmor is enabled. Should I report here or create an issue?

phaer commented:
@catuhana Thanks for reporting: 
Ideally an issue, referencing this PR
